nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/TestSQLRepositoryQuery.java

```diff
@@ -2702,6 +2717,55 @@ public void testQueryComplexWhere() throws Exception {
         assertEquals(Arrays.asList(docId), getIds(res));
     }
 
+    @Test
+    public void testQueryComplexCorrelation() throws Exception {
+        DocumentModel doc = session.createDocumentModel("/", "doc", "TestDoc");
+        doc.setPropertyValue("tst:title", "title");
+        Map<String, Object> friend0 = new HashMap<>();
+        friend0.put("firstname", "John");
+        friend0.put("lastname", "Lennon");
+        Map<String, Object> friend1 = new HashMap<>();
+        friend1.put("firstname", "Adam");
+        friend1.put("lastname", "Smith");
+        doc.setPropertyValue("tst:friends", (Serializable) Arrays.asList(friend0, friend1));
+        doc = session.createDocument(doc);
+        session.save();
+
+        String clause;
+        DocumentModelList res;
+
+        // query with correlation, and no single element of the map matches the query
+        // the backend must interpret the correlation correctly
+
+        clause = "tst:friends/*1/firstname = 'John' AND " //
+                + "tst:friends/*1/lastname = 'Smith'";
+        res = session.query(SELECT_WHERE + clause);
+        assertEquals(0, res.size());
+
+        // same one with interleaved unrelated clause
+
+        clause = "tst:friends/*1/firstname = 'John' AND " //
+                + "tst:title = 'foo' AND " //
+                + "tst:friends/*1/lastname = 'Smith'";
+        res = session.query(SELECT_WHERE + clause);
+        assertEquals(0, res.size());
+
+        // more complex correlation with sub-clauses
+
+        clause = "(tst:friends/*1/firstname = 'John' OR tst:friends/*1/firstname = 'XYZ') AND " //
+                + "(tst:friends/*1/lastname = 'Smith' OR tst:friends/*1/lastname = 'XYZ')";
+        res = session.query(SELECT_WHERE + clause);
+        assertEquals(0, res.size());
+
+        // same one with interleaved unrelated clause
+
+        clause = "(tst:friends/*1/firstname = 'John' OR tst:friends/*1/firstname = 'XYZ') AND " //
+                + "tst:title = 'foo' AND" //
+                + "(tst:friends/*1/lastname = 'Smith' OR tst:friends/*1/lastname = 'XYZ')";
+        res = session.query(SELECT_WHERE + clause);
+        assertEquals(0, res.size());
+    }
```

Comment:
May be add a test that match so we are sure the document are searchable

Reply:
Good idea, done